### PR TITLE
feat: add tracing to multichain account service

### PR DESCRIPTION
**Pull request by montelaidev**
## Explanation

This PR adds traces to the Multichain Account Service. 

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've communicated my changes to consumers by [updating changelogs for packages I've changed](https://github.com/MetaMask/core/tree/main/docs/contributing.md#updating-changelogs), highlighting breaking changes as necessary
- [ ] I've prepared draft pull requests for clients and consumer packages to resolve any breaking changes

---

> [!NOTE]
> Adds pluggable tracing to account discovery across EVM, Solana, Bitcoin, and Tron providers with a default no-op fallback and service-level configuration.
> 
> - **Tracing/Analytics**:
>   - Add `TraceName` enum and `traceFallback` in `src/analytics/*` with exports.
>   - Extend `SnapAccountProvider` to accept a `TraceCallback` and provide a protected `trace` wrapper.
>   - Wrap `discoverAccounts` in tracing for `EvmAccountProvider`, `SolAccountProvider`, `TrxAccountProvider`, and `BtcAccountProvider` (defaulting to `traceFallback`).
>   - `MultichainAccountService` accepts optional `config.trace` and passes it to providers.
> - **Types**:
>   - Add `MultichainAccountServiceConfig` with optional `trace` and plumb through constructors.
> - **Tests**:
>   - Add comprehensive trace behavior tests for all providers and `SnapAccountProvider` (custom vs fallback, empty/error cases).
>   - Update service/provider constructor tests to assert `TraceCallback` is passed; minor event emission assertion tightened.
> - **Docs/Meta**:
>   - Update `CHANGELOG.md` noting optional tracing configuration.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit aebc20c00cf504df5c476c27a5df9f6c2599c0b5. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/multichain-account-service/src/MultichainAccountWallet.ts [hunk 1]
```diff
@@ -338,12 +344,22 @@ export class MultichainAccountWallet<
...
-          this.#providers.map((provider) =>
-            provider.createAccounts({
-              entropySource: this.#entropySource,
-              groupIndex,
-            }),
-          ),
+          this.#providers.map(async (provider) => {
+            return await this.#trace(
+              {
+                name: TraceName.SnapDiscoverAccounts,
```
Comment (ccharly) [inline packages/multichain-account-service/src/MultichainAccountWallet.ts]: We're not running any discovery here, so we might want to add a new trace constant for this

File: packages/multichain-account-service/src/MultichainAccountWallet.ts [hunk 2]
```diff
@@ -388,18 +404,36 @@ export class MultichainAccountWallet<
...
-              // Log errors from background providers but don't fail the operation
-              const errorMessage = `Could not to create account with provider "${provider.getName()}" for multichain account group index: ${groupIndex}`;
-              this.#log(`${WARNING_PREFIX} ${errorMessage}:`, error);
-            });
-        });
+        for (const provider of otherProviders) {
+          (async () => {
+            await this.#trace(
+              {
+                name: TraceName.SnapDiscoverAccounts,
```
Comment (ccharly) [inline packages/multichain-account-service/src/MultichainAccountWallet.ts]: Same here.

File: packages/multichain-account-service/src/MultichainAccountWallet.test.ts [hunk 3]
```diff
@@ -341,6 +386,90 @@ describe('MultichainAccountWallet', () => {
...
+        .withGroupIndex(groupIndex)
+        .get();
+
+      evmProvider.createAccounts.mockResolvedValueOnce([mockNextEvmAccount]);
+      solProvider.createAccounts.mockResolvedValueOnce([]);
+
+      await wallet.createMultichainAccountGroup(groupIndex);
+
+      // Wait for background operations to complete
+      await new Promise((resolve) => setTimeout(resolve, 10));
```
Comment (ccharly) [inline packages/multichain-account-service/src/MultichainAccountWallet.test.ts]: Could you introduce this helper:
- https://github.com/MetaMask/core/blob/7ac6585651c8c6752928f5124b93edce611a24bd/packages/multichain-transactions-controller/src/MultichainTransactionsController.test.ts#L248-L253

And use it here after?

File: packages/multichain-account-service/src/MultichainAccountWallet.ts [hunk 4]
```diff
@@ -70,17 +74,19 @@ export class MultichainAccountWallet<
...
     messenger: MultichainAccountServiceMessenger;
+    trace?: TraceCallback;
   }) {
     this.#id = toMultichainAccountWalletId(entropySource);
     this.#providers = providers;
     this.#entropySource = entropySource;
     this.#messenger = messenger;
     this.#accountGroups = new Map();
-
+    this.#trace = trace ?? (((_request, fn) => fn?.()) as TraceCallback);
```
Comment (ccharly) [inline packages/multichain-account-service/src/MultichainAccountWallet.ts]: Would be nice to add a `traceFallback` function somewhere, so we can re-use it as fallback for `MultichainAccount{Wallet,Service}`!

(That's what we have for B&S: https://github.com/MetaMask/core/blob/main/packages/account-tree-controller/src/backup-and-sync/analytics/traces.ts#L11-L29)

File: packages/multichain-account-service/src/constants/traces.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export enum TraceName {
+  'SnapDiscoverAccounts' = 'Snap Discover Accounts',
```
Comment (ccharly) [inline packages/multichain-account-service/src/constants/traces.ts:2]: In this service we also handle native keyring like for EVM, so we should not mention "Snap" here IMO.

File: packages/multichain-account-service/src/MultichainAccountWallet.ts (hunk 1 above)
Comment (montelaidev) [reply packages/multichain-account-service/src/MultichainAccountWallet.ts]: This has been moved to the wrapper now.

File: packages/multichain-account-service/src/MultichainAccountWallet.ts (hunk 2 above)
Comment (montelaidev) [reply packages/multichain-account-service/src/MultichainAccountWallet.ts]: Moved to the wrapper now.

File: packages/multichain-account-service/src/MultichainAccountWallet.test.ts (hunk 3 above)
Comment (montelaidev) [reply packages/multichain-account-service/src/MultichainAccountWallet.test.ts]: Moved to the wrapper now.

File: packages/multichain-account-service/src/constants/traces.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
```
Comment (ccharly) [inline packages/multichain-account-service/src/constants/traces.ts:2]: I still think `Snap*` is too generic. Since we're providing the `providerName` in the trace parameter, maybe we can just use `DiscoverAccounts` for any providers instead? And we always pass a `providerName` parameter?

File: packages/multichain-account-service/src/providers/EvmAccountProvider.ts [hunk 7]
```diff
@@ -213,38 +220,48 @@ export class EvmAccountProvider extends BaseBip44AccountProvider {
     entropySource: EntropySourceId;
     groupIndex: number;
   }): Promise<Bip44Account<KeyringAccount>[]> {
-    const provider = this.getEvmProvider();
-    const { entropySource, groupIndex } = opts;
+    return this.#trace(
+      {
+        name: TraceName.EvmDiscoverAccounts,
+        data: {
+          providerName: this.getName(),
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/EvmAccountProvider.ts]: I'd just rename `providerName` to `provider` in this context

File: packages/multichain-account-service/src/providers/EvmAccountProvider.test.ts [hunk 8]
```diff
@@ -408,4 +408,117 @@ describe('EvmAccountProvider', () => {
...
+    });
+
+    expect(result).toStrictEqual([expectedAccount]);
+  });
+
+  it('trace callback is called even when discovery returns empty results', async () => {
+    const mockTrace = jest.fn().mockImplementation(async (request, fn) => {
+      expect(request.name).toBe('EVM Discover Accounts');
+      expect(request.data).toStrictEqual({
+        providerName: 'EVM',
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/EvmAccountProvider.test.ts]: let's use the constant `EVM_ACCOUNT_PROVIDER_NAME` here

File: packages/multichain-account-service/src/providers/EvmAccountProvider.test.ts [hunk 9]
```diff
@@ -408,4 +408,117 @@ describe('EvmAccountProvider', () => {
       }),
     ).toStrictEqual([MOCK_HD_ACCOUNT_1]);
   });
+
+  it('calls trace callback during account discovery', async () => {
+    const mockTrace = jest.fn().mockImplementation(async (request, fn) => {
+      expect(request.name).toBe('EVM Discover Accounts');
+      expect(request.data).toStrictEqual({
+        providerName: 'EVM',
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/EvmAccountProvider.test.ts]: Same, use the constant here too

File: packages/multichain-account-service/CHANGELOG.md [hunk 10]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+
+- Adds an optional trace argument to MultichainAccountService and MultichainAccountWallet ([#7006](https://github.com/MetaMask/core/pull/7006))
```
Comment (ccharly) [inline packages/multichain-account-service/CHANGELOG.md]:
```suggestion
### Added

- Add optional tracing configuration ([#7006](https://github.com/MetaMask/core/pull/7006))
  - For now, only the account discovery is being traced.
```

File: packages/multichain-account-service/src/MultichainAccountService.ts [hunk 11]
```diff
@@ -81,28 +89,32 @@ export class MultichainAccountService {
...
         this.#messenger,
         providerConfigs?.[EvmAccountProvider.NAME],
+        this.#trace.bind(this),
       ),
       new AccountProviderWrapper(
         this.#messenger,
         new SolAccountProvider(
           this.#messenger,
           providerConfigs?.[SolAccountProvider.NAME],
+          this.#trace.bind(this),
```
[the remark this answers is not in the document]
Comment (ccharly) [reply packages/multichain-account-service/src/MultichainAccountService.ts]: I think Cursor is right here. We should probably do this instead:

```ts
const trace = config?.trace ?? traceFallback;

this.#messenger = messenger;
...
this.#trace = trace;

...
    this.#providers = [
      new EvmAccountProvider(
        this.#messenger,
        providerConfigs?.[EvmAccountProvider.NAME],
        trace,
      ),
    ...
```

File: packages/multichain-account-service/src/constants/traces.ts (hunk 6 above)
Comment (ccharly) [reply packages/multichain-account-service/src/constants/traces.ts:2]: @montelaidev I think my comment still stands, maybe we can just use `DiscoverAccounts` and just use the `provider` key passed alongside this trace to identify which kind of discovery has ran?

File: packages/multichain-account-service/src/providers/SolAccountProvider.test.ts [hunk 12]
```diff
@@ -113,6 +114,15 @@ function setup({
     .mockImplementation((address: string) =>
       keyring.accounts.find((account) => account.address === address),
     );
+
+  const mockTrace = jest.fn().mockImplementation(async (request, fn) => {
+    expect(request.name).toBe('Snap Discover Accounts');
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/SolAccountProvider.test.ts]: We probably wanna re-use `TraceName` constants here?

File: packages/multichain-account-service/src/providers/SolAccountProvider.test.ts [hunk 13]
```diff
@@ -113,6 +114,15 @@ function setup({
...
+    expect(request.data).toStrictEqual({
+      provider: 'Solana',
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/SolAccountProvider.test.ts]: We can also re-use the account provider name constant here

File: packages/multichain-account-service/src/providers/TrxAccountProvider.test.ts [hunk 14]
```diff
@@ -308,4 +308,133 @@ describe('TrxAccountProvider', () => {
...
+      expect(discovered).toHaveLength(1);
+      // No trace errors, fallback trace should be used silently
+    });
+
+    it('trace callback is called even when discovery returns empty results', async () => {
+      const mockTrace = jest.fn().mockImplementation(async (request, fn) => {
+        expect(request.name).toBe('Snap Discover Accounts');
+        expect(request.data).toStrictEqual({
+          provider: 'Tron',
+        });
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/TrxAccountProvider.test.ts]: Same comment than for Solana

File: packages/multichain-account-service/src/providers/SnapAccountProvider.test.ts [hunk 15]
```diff
@@ -1,18 +1,34 @@
...
 
+jest.mock('../analytics', () => ({
+  ...jest.requireActual('../analytics'),
+  traceFallback: jest
+    .fn()
+    .mockImplementation(async (_request: TraceRequest, fn?: () => unknown) => {
+      if (fn) {
+        return await fn();
+      }
+      return undefined;
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/SnapAccountProvider.test.ts]: We can just re-use `traceFallback` here maybe? Or do we really need to match that at all? 🤔

File: packages/multichain-account-service/src/providers/BtcAccountProvider.test.ts [hunk 16]
```diff
@@ -324,4 +324,129 @@ describe('BtcAccountProvider', () => {
...
+      expect(discovered).toHaveLength(1);
+      // No trace errors, fallback trace should be used silently
+    });
+
+    it('trace callback is called even when discovery returns empty results', async () => {
+      const mockTrace = jest.fn().mockImplementation(async (request, fn) => {
+        expect(request.name).toBe('Snap Discover Accounts');
+        expect(request.data).toStrictEqual({
+          provider: 'Bitcoin',
+        });
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/BtcAccountProvider.test.ts]: We can re-use constants here too

File: packages/multichain-account-service/src/providers/BtcAccountProvider.test.ts [hunk 17]
```diff
@@ -324,4 +324,129 @@ describe('BtcAccountProvider', () => {
...
     expect(discovered).toStrictEqual([]);
   });
+
+  describe('trace functionality', () => {
+    it('calls trace callback during account discovery', async () => {
+      const mockTrace = jest.fn().mockImplementation(async (request, fn) => {
+        expect(request.name).toBe('Snap Discover Accounts');
+        expect(request.data).toStrictEqual({
+          provider: 'Bitcoin',
+        });
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/BtcAccountProvider.test.ts]: We can re-use constants here too

File: packages/multichain-account-service/src/providers/SnapAccountProvider.test.ts [hunk 18]
```diff
@@ -128,6 +248,152 @@ describe('SnapAccountProvider', () => {
...
+      mockTraceCallback.mockImplementation(async (_request, fn) => fn?.());
+
+      // Test with only messenger parameter (uses default config and trace)
+      const solProvider = new SolAccountProvider(mockMessenger);
+      const request = { name: 'Test Request', data: {} };
+      const fn = jest.fn().mockResolvedValue('defaultResult');
+
+      // Access protected trace method
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      await (solProvider as any).trace(request, fn);
```
Comment (ccharly) [inline packages/multichain-account-service/src/providers/SnapAccountProvider.test.ts]: That seems pretty tricky IMO...

And another solution would be to create a simple `MockAccountProvider` class and inherits from `SnapAccountProvider`. Then we can have a "fake" method triggering the protected `trace` method.

Having this `MockAccountProvider` would also allow us to test all the various constructor combinations IMO (instead of relying on the `SolAccountProvider`)

WDYT?

File: packages/multichain-account-service/src/analytics/traces.test.ts [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+describe('MultichainAccountService - Traces', () => {
+  describe('TraceName', () => {
+    it('contains expected trace names', () => {
+      expect(TraceName).toStrictEqual({
+        SnapDiscoverAccounts: 'Snap Discover Accounts',
+        EvmDiscoverAccounts: 'EVM Discover Accounts',
+      });
+    });
+  });
+
```
Comment (ccharly) [inline packages/multichain-account-service/src/analytics/traces.test.ts:15]: Is this required for coverage purposes? 😮

Comment (montelaidev) [reply packages/multichain-account-service/src/analytics/traces.test.ts:15]: Yea for coverage :(
